js/marshal.go

```diff
@@ -0,0 +1,224 @@
+package js
+
+import (
+	"syscall/js"
+)
+
+type (
+	Type  int
+	Value struct {
+		_type        Type
+		_constructor string
+
+		value interface{}
+	}
+)
+
+const (
+	TypeNumber Type = iota
+	TypeBoolean
+	TypeString
+	TypeObject
+	TypeArray
+	TypeNull
+)
+
+func NewValue(value interface{}) *Value {
+	switch value.(type) {
+	case int, int32, int64, uint, uint32, uint64, float32, float64:
+		return &Value{
+			_type:        TypeNumber,
+			_constructor: "Number",
+			value:        value,
+		}
+	case bool:
+		return &Value{
+			_type:        TypeBoolean,
+			_constructor: "Boolean",
+			value:        value,
+		}
+	case string:
+		return &Value{
+			_type:        TypeString,
+			_constructor: "String",
+			value:        value,
+		}
+	case map[string]interface{}:
+		return &Value{
+			_type:        TypeObject,
+			_constructor: "Object",
+			value:        value,
+		}
+	case []interface{}:
+		return &Value{
+			_type:        TypeArray,
+			_constructor: "Array",
+			value:        value,
+		}
+	default:
+		return &Value{
+			_type:        TypeNull,
+			_constructor: "Null",
+			value:        nil,
+		}
+	}
+}
+
+func (v Value) Value() interface{} {
+	return v.value
+}
+
+func (v Value) Get(key string) interface{} {
+	val, ok := v.value.(map[string]interface{})[key]
+	if !ok {
+		return nil
+	}
+	return val.(Value).value
+}
+
+func (v Value) Set(key string, value interface{}) {
```

Comment:
What happens if?

```
myInt : = 99
newVal := js.NewValue(myInt) // that is I pass in an int or other non-object
newVal.Set(<any-key-string>, <any-other-value>)
```

Reply:
It'll fail. But that's a good one, instead of panicking, we could fail silently by returning the error to the caller and allowing them to handle it

Reply:
Actually, this is not the case. It type asserts and properly sets the value as an `interface{}` type